Src/IronPython/Runtime/MemoryView.cs

```diff
@@ -851,20 +852,70 @@ public sealed class MemoryView : ICodeFormattable, IWeakReferenceable, IBufferPr
             }
 
             TypecodeOps.DecomposeTypecode(_format, out _, out char typecode);
-            if (!TypecodeOps.IsByteCode(typecode)) {
+            if (!TypecodeOps.IsByteOrCharCode(typecode)) {
                 throw PythonOps.ValueError("memoryview: hashing is restricted to formats 'B', 'b' or 'c'");
             }
 
             _storedHash = tobytes().GetHashCode();
             return _storedHash.Value;
         }
 
+        private bool EquivalentShape(MemoryView mv) {
+            if (_numDims != mv._numDims) return false;
+            for (int i = 0; i < _numDims; i++) {
+                if (_shape[i] != mv._shape[i]) return false;
+                if (_shape[i] == 0) break;
+            }
+            return true;
+        }
+
         public bool __eq__(CodeContext/*!*/ context, [NotNone] MemoryView value) {
-            if (_buffer == null) {
-                return value._buffer == null;
+            if (_buffer == null) return ReferenceEquals(this, value);
+            if (value._buffer == null) return false;
+            if (!EquivalentShape(value)) return false;
+
+            TypecodeOps.DecomposeTypecode(_format, out char ourByteorder, out char ourTypecode);
+            // TODO: Support non-native byteorder
+
+            // fast tracks if item formats match
+            if (_format == value._format && !TypecodeOps.IsFloatCode(ourTypecode)) {
+                if (ReferenceEquals(this, value)) return true;
+
+                if (_isCContig && value._isCContig) {
+                    // compare blobs
+                    return ((IPythonBuffer)this).AsReadOnlySpan().SequenceEqual(((IPythonBuffer)value).AsReadOnlySpan());
+                }
+
+                // compare byte by byte
+                using var ourBytes = this.EnumerateBytes();
+                using var theirBytes = value.EnumerateBytes();
+                while (ourBytes.MoveNext() && theirBytes.MoveNext()) {
+                    if (ourBytes.Current != theirBytes.Current) return false;
+                }
+
+                return true;
+            }
+
+            // compare item by item
+            TypecodeOps.DecomposeTypecode(value._format, out char theirByteorder, out char theirTypecode);
+
+            if ((ourTypecode is 'c') && TypecodeOps.IsByteCode(theirTypecode) ||
+                (theirTypecode is 'c') && TypecodeOps.IsByteCode(ourTypecode)) {
+                if (context.LanguageContext.PythonOptions.BytesWarning != Microsoft.Scripting.Severity.Ignore) {
+                    PythonOps.Warn(context, PythonExceptions.BytesWarning, "Comparison between bytes and int");
```

Comment:
I think this warning may be unnecessary? It's only since 3.5 and should be triggered in the 3.6 branch by:
https://github.com/IronLanguages/ironpython3/blob/26767ee5e7de1b361acf75acca55715930476640/Src/IronPython/Runtime/Bytes.cs#L1189-L1194

Reply:
and I guess this also triggers incorrectly (at least compared to CPython) on an empty buffer...
```py
memoryview(b'') == memoryview(b'').cast('c')
```

Reply:
Oh, you have it already in 3.6. How about I revert the last commit here and save the test patch to apply once 3.6 is merged into master? Or I leave the test in since it is only run with the `-b` option which is not enabled by default in standard test runs?

Reply:
On the other hand, the warning from 3.6 will not be triggered when hitting one of the fast tracks in `memoryview.__eq__`. (To be honest, neither my latest update would do it but this can easily be fixed).

Reply:
> Oh, you have it already in 3.6. How about I revert the last commit here and save the test patch to apply once 3.6 is merged into master? Or I leave the test in since it is only run with the `-b` option which is not enabled by default in standard test runs?

I guess you could leave the test in and remove the `is_cli` bit, will automatically work in 3.6 since the version check will pass.

> On the other hand, the warning from 3.6 will not be triggered when hitting one of the fast tracks in `memoryview.__eq__`. (To be honest, neither my latest update would do it but this can easily be fixed).

Doesn't the fast track only compare same formats so there should never be a warning?

Reply:
> Doesn't the fast track only compare same formats so there should never be a warning?

You are right, my mistake.